src/Analyzer.Reports/SarifReportWriter.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using Microsoft.Azure.Templates.Analyzer.Types;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Writers;
+
+namespace Microsoft.Azure.Templates.Analyzer.Reports
+{
+    /// <summary>
+    /// Class to export analysis result to SARIF report
+    /// </summary>
+    public class SarifReportWriter : IReportWriter
+    {
+        internal const string UriBaseIdString = "ROOTPATH";
+        internal const string PeriodString = ".";
+
+        private IFileInfo reportFile;
+        private Run sarifRun;
+        private IList<Result> sarifResults;
+        private IDictionary<string, ReportingDescriptor> rulesDictionary;
+        private string rootPath;
+
+        /// <summary>
+        /// Constructor of the SarifReportWriter class
+        /// </summary>
+        /// <param name="reportFile">File where the report will be written</param>
+        /// <param name="targetPath">The directory that will be analyzed</param>
+        public SarifReportWriter(IFileInfo reportFile, string targetPath = null)
+        {
+            this.reportFile = reportFile ?? throw new ArgumentException(nameof(reportFile));
+            this.InitRun();
+            this.rulesDictionary = new ConcurrentDictionary<string, ReportingDescriptor>();
+            this.sarifResults = new List<Result>();
+            this.rootPath = targetPath;
+        }
+
+        /// <inheritdoc/>
+        public void WriteResults(IEnumerable<IEvaluation> evaluations, IFileInfo templateFile, IFileInfo parameterFile = null)
+        {
+            this.rootPath ??= templateFile.DirectoryName;
+            foreach (var evaluation in evaluations.Where(eva => !eva.Passed))
+            {
+                // get rule definition from first level evaluation
+                ReportingDescriptor rule = this.ExtractRule(evaluation);
+                this.ExtractResult(evaluation, evaluation, templateFile.FullName);
+            }
+        }
+
+        internal Run SarifRun => this.sarifRun;
+
+        private void InitRun()
+        {
+            this.sarifRun = new Run
+            {
+                Tool = new Tool
+                {
+                    Driver = new ToolComponent
+                    {
+                        Name = Constants.ToolName,
+                        FullName = Constants.ToolFullName,
+                        Version = Constants.ToolVersion,
+                        InformationUri = new Uri(Constants.InformationUri),
+                        Organization = Constants.Organization,
+                    }
+                }
+            };
+        }
+
+        private ReportingDescriptor ExtractRule(IEvaluation evaluation)
+        {
+            if (!rulesDictionary.TryGetValue(evaluation.RuleId, out _))
```

Comment:
To avoid double-lookup (in the return statement at the end), instead of discarding the result here, save the result, assign `new ReportingDescriptor()` to it when creating it below (and add that to the dictionary), then return the variable at the end.

Reply:
Fixed